x509-cert/src/serial_number.rs

```diff
@@ -32,15 +32,27 @@ impl SerialNumber {
     /// Maximum length in bytes for a [`SerialNumber`]
     pub const MAX_LEN: Length = Length::new(20);
 
+    /// See notes in `SerialNumber::new` and `SerialNumber::decode_value`.
+    const MAX_DECODE_LEN: Length = Length::new(21);
+
     /// Create a new [`SerialNumber`] from a byte slice.
+    ///
+    /// The byte slice **must** represent a positive integer.
     pub fn new(bytes: &[u8]) -> Result<Self> {
-        let inner = Int::new(bytes)?;
-
-        if inner.len() > SerialNumber::MAX_LEN {
+        let inner = Uint::new(bytes)?;
+
+        // The user might give us a 20 byte unsigned integer with a high MSB,
+        // which we'd then encode with 21 octets to preserve the sign bit.
+        // RFC 5280 is ambiguous about whether this is valid, so we limit
+        // `SerialNumber` *encodings* tp 20 bytes or fewer while permitting
+        // `SerialNumber` *decodings* to have up to 21 bytes below.
+        if inner.value_len()? > SerialNumber::MAX_LEN {
             return Err(ErrorKind::Overlength.into());
         }
 
-        Ok(Self { inner })
+        Ok(Self {
+            inner: inner.into(),
+        })
     }
```

Comment:
This ensures that we never create a *new* negative serial number, but the way it does that is potentially surprising because of how the `Int::new` API (unchanged) works. In particular, `Int::new` *assumes* that its input is positive, and encodes it correctly to enforce this assumption.

I've updated the tests below to show this behavior.

Reply:
(This arguably violates POLA, since we're changing the serial number given to us by the user. But it's how the current `Int::new` API behaves, and I didn't want to make changes to that before putting something initial up here :slightly_smiling_face:)

Reply:
curious: what does POLA stands for?

Reply:
Principle of least astonishment

Reply:
Yep! Sorry for the acronym abuse.

By POLA in this case I meant that users might consider it astonishing that the serial they provide doesn't end up in the cert, even if the *result* is technically more correct.